onnxruntime/contrib_ops/cpu/signal/dft.cc

```diff
@@ -0,0 +1,521 @@
+         (BitReverseTable256[(num >> 24) & 0xff]);
+  return static_cast<uint32_t>(((uint64_t)rev) >> (32 - TSignificantBits));
+}
+
+template <typename T>
+static inline T bit_reverse(T num, unsigned significant_bits) {
+  switch (significant_bits) {
+    case 0: return static_cast<T>(bit_reverse<0>(static_cast<uint32_t>(num)));
+    case 1: return static_cast<T>(bit_reverse<1>(static_cast<uint32_t>(num)));
+    case 2: return static_cast<T>(bit_reverse<2>(static_cast<uint32_t>(num)));
+    case 3: return static_cast<T>(bit_reverse<3>(static_cast<uint32_t>(num)));
+    case 4: return static_cast<T>(bit_reverse<4>(static_cast<uint32_t>(num)));
+    case 5: return static_cast<T>(bit_reverse<5>(static_cast<uint32_t>(num)));
+    case 6: return static_cast<T>(bit_reverse<6>(static_cast<uint32_t>(num)));
+    case 7: return static_cast<T>(bit_reverse<7>(static_cast<uint32_t>(num)));
+    case 8: return static_cast<T>(bit_reverse<8>(static_cast<uint32_t>(num)));
+    case 9: return static_cast<T>(bit_reverse<9>(static_cast<uint32_t>(num)));
+    case 10: return static_cast<T>(bit_reverse<10>(static_cast<uint32_t>(num)));
+    case 11: return static_cast<T>(bit_reverse<11>(static_cast<uint32_t>(num)));
+    case 12: return static_cast<T>(bit_reverse<12>(static_cast<uint32_t>(num)));
+    case 13: return static_cast<T>(bit_reverse<13>(static_cast<uint32_t>(num)));
+    case 14: return static_cast<T>(bit_reverse<14>(static_cast<uint32_t>(num)));
+    case 15: return static_cast<T>(bit_reverse<15>(static_cast<uint32_t>(num)));
+    case 16: return static_cast<T>(bit_reverse<16>(static_cast<uint32_t>(num)));
+    case 17: return static_cast<T>(bit_reverse<17>(static_cast<uint32_t>(num)));
+    case 18: return static_cast<T>(bit_reverse<18>(static_cast<uint32_t>(num)));
+    case 19: return static_cast<T>(bit_reverse<19>(static_cast<uint32_t>(num)));
+    case 20: return static_cast<T>(bit_reverse<20>(static_cast<uint32_t>(num)));
+    case 21: return static_cast<T>(bit_reverse<21>(static_cast<uint32_t>(num)));
+    case 22: return static_cast<T>(bit_reverse<22>(static_cast<uint32_t>(num)));
+    case 23: return static_cast<T>(bit_reverse<23>(static_cast<uint32_t>(num)));
+    case 24: return static_cast<T>(bit_reverse<24>(static_cast<uint32_t>(num)));
+    case 25: return static_cast<T>(bit_reverse<25>(static_cast<uint32_t>(num)));
+    case 26: return static_cast<T>(bit_reverse<26>(static_cast<uint32_t>(num)));
+    case 27: return static_cast<T>(bit_reverse<27>(static_cast<uint32_t>(num)));
+    case 28: return static_cast<T>(bit_reverse<28>(static_cast<uint32_t>(num)));
+    case 29: return static_cast<T>(bit_reverse<29>(static_cast<uint32_t>(num)));
+    case 30: return static_cast<T>(bit_reverse<30>(static_cast<uint32_t>(num)));
+    case 31: return static_cast<T>(bit_reverse<31>(static_cast<uint32_t>(num)));
+    case 32: return static_cast<T>(bit_reverse<32>(static_cast<uint32_t>(num)));
+    default: ORT_THROW("Unsupported bit size.");
+  }
+}
```

Comment:
Using compiler intrinsics might be worth it

Reply:
Which intrinsic?

Reply:
popcnt for population count (number of 1 bits), rbit to reverse bits in ARM. For bit reverse in x86 there's fancy ways using SIMD that are reportedly >5 times faster than the lookup table solution, like https://stackoverflow.com/a/24058332/1390418 and https://stackoverflow.com/a/46318399/1390418, but it's much trickier to implement